[AUTOMATIONS-CONTRACT] Vesting Scheduler V3

0xPilou opened the pull request:
Vesting Scheduler V3 offers the possibility for vesting sender to update their vesting schedule amounts and/or durations.

_Modifications includes :_ 
- Add total amount and already vested amount tracking to VestingSchedule
- Add overloaded `updateVestingSchedule` method with dynamic flow rate calculation
- Update interface to include new vesting schedule tracking fields
- Review early termination and compensation logic

_Optimism Sepolia Deployment :_ 
Contract is deployed at [0x2dCFbF5BcE0522257E78bF164871770D30634A96](https://sepolia-optimism.etherscan.io/address/0x2dCFbF5BcE0522257E78bF164871770D30634A96#code)

0xPilou committed:
chore: init SchedulerV3 contract & interface

0xPilou committed:
feat: Enhance vesting schedule tracking and update mechanism

0xPilou committed:
refactor: Rename updateVestingSchedule method to updateVestingScheduleAmount

0xPilou committed:
test: draft comprehensive test suite for VestingSchedulerV3

0xPilou committed:
refactor: Improve VestingSchedulerV3 schedule update logic

- Rename `updateVestingSchedule` to `updateVestingScheduleEndDate`
- Optimize schedule update mechanism with more precise vesting calculations
- Simplify flow rate and remainder amount handling
- Update error handling and edge case management

0xPilou committed:
test: Enhance VestingSchedulerV3 test coverage for schedule updates

- Add comprehensive test cases for `updateVestingScheduleEndDate` and `updateVestingScheduleAmount`
- Improve error handling tests for non-existent vesting schedules
- Refactor existing tests to use new method signatures
- Optimize test scenarios for end date and amount updates

d10r committed:
added support for batch calls using OPERATION_TYPE_ERC2771_FORWARD_CALL

0xPilou committed:
refactor: Optimize VestingSchedulerV3 settlement and tracking mechanisms

- Introduce `_settle` method to handle vesting schedule updates and tracking
- Remove `totalAmount` from storage and improve vesting calculation logic
- Add `lastUpdated` timestamp to track precise vesting progress
- Enhance early termination and compensation handling
- Commented out unused update methods with TODOs for future implementation

0xPilou committed:
test: Refine VestingSchedulerV3 settlement and testing logic

- Update test cases to reflect recent refactoring of vesting schedule mechanisms
- Remove commented-out test methods for non-existent schedule updates
- Improve settlement calculation by using total amount to vest
- Simplify vesting schedule verification in test scenarios

0xPilou committed:
fix: Correct vesting amount calculation and settlement logic

- Adjust settlement calculation to subtract cliff amount from already vested amount
- Optimize total vested amount calculation by improving parentheses placement
- Refactor `_executeCliffAndFlow` and `executeEndVesting` to handle vesting amounts more precisely

0xPilou committed:
test: Refactor VestingSchedulerV3 test scenarios for precise settlement

- Update test cases to improve vesting amount and timing calculations
- Simplify test setup and execution flow
- Enhance precision in settlement and balance verification
- Optimize time warping and event expectation in test methods

0xPilou committed:
fix: Improve vesting amount calculation and remainder tracking

- Refactor `_getTotalVestedAmount` to enhance clarity and precision of vesting calculations
- Correct remainder amount calculation by subtracting flow rate duration from total vested amount
- Add comments to clarify vesting logic for different schedule stages
- Optimize flow amount calculation with improved variable naming

0xPilou committed:
test: Refactor early vesting end test with improved randomization

- Replace hardcoded early end time calculation with bounded randomization
- Remove unused total amount to vest calculation
- Comment out previous early end time calculation method
- Improve test scenario randomness and flexibility

0xPilou committed:
test: fallback to previous early vesting end time calculation method

- Restore original early end time calculation using division
- Remove bounded randomization approach
- Simplify early vesting end time selection logic

0xPilou committed:
feat: Implement updateVestingScheduleFlowRateFromAmount and refine schedule update logic

- Add implementation for `updateVestingScheduleFlowRateFromAmount` method
- Enhance schedule update validation with comprehensive time window checks
- Improve settlement logic in `_settle` method
- Optimize flow rate and remainder amount calculations during schedule updates
- Add comments to clarify update conditions and vesting logic

0xPilou committed:
feat: Add detailed events for vesting schedule updates in VestingSchedulerV3

- Implement new events `VestingScheduleTotalAmountUpdated` and `VestingScheduleEndDateUpdated`
- Update event emissions to include remainder amount and flow rate details
- Rename `updateVestingScheduleEndDate` to `updateVestingScheduleFlowRateFromEndDate`
- Enhance event logging with comprehensive schedule update information

0xPilou committed:
test: Update VestingSchedulerV3 test cases for flow rate and end date updates

- Modify test methods to use `updateVestingScheduleFlowRateFromEndDate`
- Enhance calculation of expected remainder and new flow rate
- Update test assertions to verify total vested amount and balance transfers
- Refactor batch operation encoding to match new method signature

0xPilou committed:
refactor: Remove commented-out updateVestingScheduleAmountAndEndDate method

- Delete unused and commented-out method from VestingSchedulerV3
- Clean up code by removing deprecated implementation with TODO comments
- Simplify contract by eliminating unimplemented method

0xPilou committed:
test: Add comprehensive test cases for VestingSchedulerV3 schedule updates

- Implement test scenarios for updating vesting schedules with longer and shorter durations
- Add test cases for updating vesting schedule amounts (both larger and smaller)
- Uncomment and update previously commented-out test method for non-existent schedule updates
- Enhance test coverage for flow rate and remainder calculations during schedule modifications

0xPilou committed:
test: added testing scenarios

0xPilou committed:
refactor: reordre test scripts

0xPilou committed:
Merge pull request #1 from d10r/add-2771-support

added support for batch calls using OPERATION_TYPE_ERC2771_FORWARD_CALL

0xPilou committed:
test: Update test method to use updateVestingScheduleFlowRateFromEndDate

0xPilou committed:
refactor: updated deploy script and .env.example to support VestingSchedulerV3 deployment

0xPilou committed:
Merge branch 'dev' into vesting-scheduler-v3

0xPilou committed:
test: increase test coverage

0xPilou committed:
refactor: comments & format

0xPilou committed:
deploy: VestingSchedulerV3 on OP Sepolia

kasparkallas commented on packages/automation-contracts/scheduler/contracts/VestingSchedulerV3.sol:
```diff
@@ -0,0 +1,815 @@
+
+        uint32 endDate = startDate + totalDuration;
+        int96 flowRate = SafeCast.toInt96(SafeCast.toInt256(totalAmount / totalDuration));
+        uint96 remainderAmount = SafeCast.toUint96(totalAmount - (SafeCast.toUint256(flowRate) * totalDuration));
```
It would be good to have this in a separate function and re-use it in the update function.

kasparkallas commented on packages/automation-contracts/scheduler/contracts/VestingSchedulerV3.sol:
```diff
@@ -0,0 +1,815 @@
+
+        // Update the vesting flow rate and remainder amount
+        vestingSchedules[agg.id].flowRate = SafeCast.toInt96(
+            SafeCast.toInt256(totalVestedAmount - alreadyVestedAmount) / SafeCast.toInt256(endDate - block.timestamp)
```
Might be more readable to have a variables:
```
uint256 amountLeftToVest = _getTotalVestedAmount(schedule) - _settleAlreadyVestedAmount(agg);
uint32 timeLeftToVest = endDate - block.timestamp;
```

And pass those into a re-used function that calculates the flow rate?

0xPilou replied:
Agreed and updated

kasparkallas commented on packages/automation-contracts/scheduler/contracts/VestingSchedulerV3.sol:
```diff
@@ -0,0 +1,815 @@
+
+    function _settle(ScheduleAggregate memory agg) internal returns (uint256 alreadyVestedAmount) {
+        VestingSchedule memory schedule = agg.schedule;
+
```
Best to have the guard clauses (reverts or asserts) in here as well, just in case?

0xPilou replied:
agreed. Added assert clause in `_settle` ensuring that settlement cannot occur before the `cliffAndFlowDate`

0xPilou committed:
refactor: rearrange calculation for better readibility

0xPilou committed:
feat: add timestamp validation for cliff and flow date in vesting settlement

0xPilou committed:
refactor: extract flow rate and remainder amount calculation into helper method

0xPilou committed:
refactor: separate accounting data from vesting schedule - review V3 interface (inherits V2 + added new functions)

0xPilou committed:
test: update test cases to use IVestingSchedulerV2 interface

0xPilou committed:
refactor: improve flow rate and remainder amount calculation methods

- Extract flow rate and remainder amount calculation into separate pure methods
- Add new `_updateVestingFlowRate` method to handle flow updates with context
- Simplify flow rate update logic in `updateVestingScheduleFlowRateFromAmount` and `updateVestingScheduleEndDate`
- Add debug event for remainder amount calculation
- Minor validation improvements for end date updates

0xPilou committed:
test: add test for updating vesting schedule end date

- Implement `testUpdateVestingSchedule` to verify end date modification
- Check that cliff and flow date is reset after update
- Validate updated end date in vesting schedule storage

0xPilou committed:
refactor: remove logging event

0xPilou committed:
feat: added overloaded `createVestingScheduleFromAmountAndDuration` for non-linear cliff schedule creation

0xPilou committed:
refactor: remove note

0xPilou committed:
test: added case for new overloaded function (non linear cliff)

0xPilou committed:
test: added case for new overloaded function (non linear cliff)

0xPilou committed:
deploy: update vestingSchedulerV3 address in network configuration

0xPilou committed:
feat: add external function `getTotalVestedAmount`

0xPilou committed:
feat: add `updateVestingScheduleFlowRateFromAmountAndEndDate` function + remove `ctx` from other update functions

0xPilou committed:
test: added test for `updateVestingScheduleFlowRateFromAmountAndEndDate`

0xPilou committed:
feat: large refactoring - remove ctx based function - reformat interface and contract - added natspec comments

kasparkallas committed:
add another test with multiple updates

kasparkallas committed:
add test w multiple updates and single claim transfer

kasparkallas committed:
refactor update functions into single core logic

kasparkallas committed:
refactor to a single update event

kasparkallas committed:
rename to settledAmount and settledDate

kasparkallas committed:
add endVestingScheduleNow function

- allows to cleanly end vesting schedule immediately
- added condition not to allow vesting schedule updates when claim period has ended
- allowed updating vesting schedule end date to current block timestamp
- made executeCliffAndFlow and executeEndVesting public instead of just external

kasparkallas committed:
add inheritdoc

kasparkallas committed:
add contract-level natspec

kasparkallas committed:
remove address from metadata

kasparkallas committed:
update metadata

d10r committed:
v3 deployments

d10r committed:
added eth-mainnet

kasparkallas committed:
Merge branch 'dev' into vesting-scheduler-v3

d10r committed:
include grapqhl in root devDependencies to fix yarn install getting stuck

kasparkallas committed:
Merge branch 'dev' into vesting-scheduler-v3

kasparkallas committed:
yarn deduplicate

kasparkallas committed:
merge dev

kasparkallas committed:
fix test

hellwolf committed:
use codecov-action 5.4.2
